### PR TITLE
feat(models,training): Shard everything

### PR DESCRIPTION
**Pull request by japols**
Improved scalability for large input grid sizes by making the whole pipeline from loading the batch to computing the loss fully sharded, never materialising the whole input/output grid in GPU memory. This also helps reduce memory usage for long rollouts at lower resolutions.

When using the config option `model.keep_batch_sharded=True` together with reading the batch in shards via `dataloader.read_group_size=hardware.num_gpus_per_model`, the batch is handed to the model already sharded across the model_comm_group and the model returns it's prediction in shards. The loss is computed locally on each shard to then get the global loss via all-reduce.

9km training step with 16 gpus per model, left: keep_batch_sharded=False, right: True

Screenshot 2025-02-04 at 20 10 58

Note that keeping the batch sharded is not supported for the Hierarchical model as it's currently broken on main. Support for hierarchical will be added in a separate PR as soon as it is fixed.

----
📚 Documentation preview 📚: https://anemoi-training--121.org.readthedocs.build/en/121/

----
📚 Documentation preview 📚: https://anemoi-graphs--121.org.readthedocs.build/en/121/

----
📚 Documentation preview 📚: https://anemoi-models--121.org.readthedocs.build/en/121/

### CODE REVIEW

File: outputs/2025-01-29/15-06-12/.hydra/config.yaml
Comment (theissenhelen) [inline outputs/2025-01-29/15-06-12/.hydra/config.yaml:1]: Are these hydra files here intentionally?

Comment (japols) [reply outputs/2025-01-29/15-06-12/.hydra/config.yaml:1]: No, thanks for noticing!

**Comment by japols** [comment]
@JPXKQX maybe you could review the changes to grid_indices.py and check if the dataloader for LAM/stretched grid still works. I made the sequence sharding identical to the one we use in the model which uses torch.tensor_split(). This makes it such that the grid shards differ in size by at most 1 resulting in better load balancing.

**Comment by japols** [comment]
@ssmmnn11 maybe you could review the changes I made to the model and how it plugs in with the GraphForecaster.

**Comment by japols** [comment]
@HCookie maybe you could review the changes to the loss functions, making them compatible with sharding.

File: models/src/anemoi/models/layers/mapper.py [hunk 1]
```diff
@@ -89,27 +91,34 @@ def pre_process(self, x, shard_shapes, model_comm_group=None) -> tuple[Tensor, T
...
 
 
 class ForwardMapperPreProcessMixin:
     """Pre-processing for Forward Mapper from data -> hidden."""
 
-    def pre_process(self, x, shard_shapes, model_comm_group=None):
-        x_src, x_dst, shapes_src, shapes_dst = super().pre_process(x, shard_shapes, model_comm_group)
-        x_src = shard_tensor(x_src, 0, shapes_src, model_comm_group)
-        x_dst = shard_tensor(x_dst, 0, shapes_dst, model_comm_group)
+    def pre_process(self, x, shard_shapes, model_comm_group=None, x_src_is_sharded=False, x_dst_is_sharded=False):
```
Comment (floriankrb) [inline models/src/anemoi/models/layers/mapper.py:118]: Shouldn't `x_src_is_sharded` and `x_dst_is_sharded` be in the `__init__` (and saved in self) instead of the `pre_process` method ?

Comment (japols) [reply models/src/anemoi/models/layers/mapper.py:118]: Currently it is decided at runtime if the input/output comes/stays sharded.

**Comment by japols** [comment]
Waiting for [PR#182](https://github.com/ecmwf/anemoi-core/pull/182) and [PR#52](https://github.com/ecmwf/anemoi-core/pull/52) to be merged.

File: models/src/anemoi/models/distributed/primitives.py [hunk 2]
```diff
@@ -92,19 +92,32 @@ def _gather(
...
-        )
-        for rank in range(comm_size)
-    ]
 
-    tensor_list[comm_rank] = input_
-    if gather_in_backward:
-        dist.all_gather(tensor_list, input_, group=group)
+    all_shards_equal_shape = all(shape == shapes[0] for shape in shapes)
+
+    if all_shards_equal_shape:  # requirement for all_gather_into_tensor (should be the case)
```
Comment (ssmmnn11) [inline models/src/anemoi/models/distributed/primitives.py]: mostly be the case? :-)

File: models/src/anemoi/models/models/encoder_processor_decoder.py [hunk 3]
```diff
@@ -163,12 +166,18 @@ def _truncate_fields(self, x, A, batch_size=None, grad_checkpoint=False):
                     out.append(self._multiply_sparse(x[i, ...], A))
         return torch.stack(out)
 
-    def _assemble_input(self, x, batch_size):
+    def _assemble_input(self, x, batch_size, grid_shard_slice=None):
+        node_attributes_data = self.node_attributes(self._graph_name_data, batch_size=batch_size)
+        if grid_shard_slice is not None:
+            node_attributes_data = node_attributes_data[
+                grid_shard_slice, :
+            ]  # TODO(Jan): shard_tensor instead for gradient ?
```
Comment (ssmmnn11) [inline models/src/anemoi/models/models/encoder_processor_decoder.py]: Have you checked that the gradients are consistent with what we had before? You can plot the trainable parameter and compare magnitude changes as an easy check.

File: models/src/anemoi/models/models/encoder_processor_decoder.py [hunk 4]
```diff
@@ -179,6 +188,8 @@ def _assemble_input(self, x, batch_size):
             # these can't be registered as buffers because ddp does not like to broadcast sparse tensors
             # hence we check that they are on the correct device ; copy should only happen in the first forward run
             if self.A_down is not None:
+                if grid_shard_slice is not None:
```
Comment (ssmmnn11) [inline models/src/anemoi/models/models/encoder_processor_decoder.py]: I think we need to change the routine here.

**Comment by cathalobrien** [comment]
Great work! To use batch sharding at inference time, we need to change the predict step to pass the grid shard slices + shapes.
The below example worked for me
```python
def predict_step(
...
         if model_comm_group is not None:
                rank=torch.distributed.get_group_rank(model_comm_group)
                world_size=torch.distributed.get_world_size(model_comm_group)
                res = x.shape[-2]
                grid_shard_shapes = [res//world_size] * world_size
                if res % world_size != 0:
                    grid_shard_shapes[-1] += res % world_size
                grid_shard_slice = slice(sum(grid_shard_shapes[:rank],sum(grid_shard_shapes[rank])))
            else:
                grid_shard_shapes=None
                grid_shard_slice=None

            y_hat = self(x, model_comm_group, grid_shard_shapes=grid_shard_shapes, grid_shard_slice=grid_shard_slice)
```
Note this version returns a sharded output from predict_step, which must be recombined on the anemoi-inference. Alternatively one could gather at the end of predict step, and have only rank 0 return the full batch.

There are more changes required on the anemoi-inference side to feed each gpu a sharded batch. I suppose we should prepare a linked anemoi-inference PR and bump the minimum versions accordingly

**Comment by ssmmnn11** [comment]
> Great work! To use batch sharding at inference time, we need to change the predict step to pass the grid shard slices + shapes. The below example worked for me
> 
> ```python
> def predict_step(
> ...
>          if model_comm_group is not None:
>                 rank=torch.distributed.get_group_rank(model_comm_group)
>                 world_size=torch.distributed.get_world_size(model_comm_group)
>                 res = x.shape[-2]
>                 grid_shard_shapes = [res//world_size] * world_size
>                 if res % world_size != 0:
>                     grid_shard_shapes[-1] += res % world_size
>                 grid_shard_slice = slice(sum(grid_shard_shapes[:rank],sum(grid_shard_shapes[rank])))
>             else:
>                 grid_shard_shapes=None
>                 grid_shard_slice=None
> 
>             y_hat = self(x, model_comm_group, grid_shard_shapes=grid_shard_shapes, grid_shard_slice=grid_shard_slice)
> ```
> 
> Note this version returns a sharded output from predict_step, which must be recombined on the anemoi-inference level. Alternatively one could gather at the end of predict step, and have only rank 0 return the full batch.
> 
> There are more changes required on the anemoi-inference side to feed each gpu a sharded batch. I suppose we should prepare a linked anemoi-inference PR and bump the minimum versions accordingly

I think best would be to gather at the end of the predict step. Also for now we could keep everything in the predict step, i.e. shard the input here before passing it to the model.

File: models/src/anemoi/models/models/encoder_processor_decoder.py [hunk 5]
```diff
@@ -163,26 +166,53 @@ def _truncate_fields(self, x, A, batch_size=None, grad_checkpoint=False):
...
+                        self._truncation_data["down"]
+                        if grid_shard_slice is None
+                        else self._truncation_data["down"][:, grid_shard_slice]
+                    )
+                    self.A_down = self._make_truncation_matrix(truncation_data)
+                self.A_down = self.A_down.to(x.device)
                 x_skip = self._truncate_fields(x_skip, self.A_down)  # to coarse resolution
             if self.A_up is not None:
-                self.A_up = self.A_up.to(x_skip.device)
+                if (
```
Comment (ssmmnn11) [inline models/src/anemoi/models/models/encoder_processor_decoder.py]: we need to shard across variables here -> should be same as seq <-> head sharding

File: models/src/anemoi/models/models/ens_encoder_processor_decoder.py [hunk 6]
```diff
@@ -58,26 +59,53 @@ def _calculate_input_dim(self, model_config):
...
 
         # these can't be registered as buffers because ddp does not like to broadcast sparse tensors
         # hence we check that they are on the correct device ; copy should only happen in the first forward run
         # todo -> parallelize this
         if self.A_down is not None:
+            if (
+                self.A_down.shape[-1] != grid_shard_size
+            ):  # reload truncation matrix for correct shard slice, this should happen only once
+                LOGGER.info(f"Reloading truncation matrix for shard slice {grid_shard_slice}")
+                truncation_data = (
```
Comment (ssmmnn11) [inline models/src/anemoi/models/models/ens_encoder_processor_decoder.py]: see encoder_processor_decoder

File: training/src/anemoi/training/distributed/ensemble.py [hunk 7]
```diff
@@ -93,15 +105,21 @@ def forward(
...
         if mgroup_:
             out = _gather(input_, dim_, shapes_, group=mgroup_)
-            return _filter(out, dim_, ndevice_, memspacing_)
+            return _filter(
+                out,
+                dim_,
+                ndevice_,
+                memspacing_,
+                offset_,
+            )  # proof of concept, backwards not working like this
```
Comment (ssmmnn11) [inline training/src/anemoi/training/distributed/ensemble.py]: is this from debugging? 

I think we can remove the gather ans split ensemble routines now (and _filter and _split)

File: training/src/anemoi/training/distributed/ensemble.py [hunk 8]
```diff
@@ -159,12 +178,12 @@ def _split(input_: Tensor, dim_: int, shapes_: tuple, group: ProcessGroup) -> Te
     return input_list[rank].contiguous(memory_format=input_format)
 
 
-def _filter(input_: Tensor, dim_: int, ndevices_: int, memspacing_: int) -> Tensor:
+def _filter(input_: Tensor, dim_: int, ndevices_: int, memspacing_: int, offset_: int = 0) -> Tensor:
```
Comment (ssmmnn11) [inline training/src/anemoi/training/distributed/ensemble.py]: do we need this? or revert back? If we don't use these anymore we could also completely remove them.

File: models/src/anemoi/models/models/encoder_processor_decoder.py [hunk 9]
```diff
@@ -143,44 +149,72 @@ def _make_truncation_matrix(self, A, data_type=torch.float32):
...
+        x_skip = x[:, -1, ...]
+        x_skip = einops.rearrange(x_skip, "batch ensemble grid vars -> (batch ensemble) grid vars")
+        x_skip = self._apply_truncation(x_skip, grid_shard_shapes, model_comm_group)
+        x_skip = einops.rearrange(x_skip, "(batch ensemble) grid vars -> batch ensemble grid vars", batch=batch_size)
+
+        # concatenate input data and node attributes
+        node_attributes_data = self.node_attributes(self._graph_name_data, batch_size=batch_size)
+        if grid_shard_shapes is not None:
+            # select grid_shard_slice of node_attributes but with proper gradient propagation in bwd
+            # the fwd pass is equivalent to node_attributes_data = node_attributes_data[grid_shard_slice, ...]
```
Comment (ssmmnn11) [inline models/src/anemoi/models/models/encoder_processor_decoder.py]: I think we can remove the comment here. sharding happens like everywhere else.

File: models/src/anemoi/models/models/ens_encoder_processor_decoder.py [hunk 10]
```diff
@@ -56,35 +57,34 @@ def _calculate_shapes_and_indices(self, data_indices: dict) -> None:
...
-        # todo -> parallelize this
-        if self.A_down is not None:
-            self.A_down = self.A_down.to(x.device)
-            x_skip = self._truncate_fields(x_skip, self.A_down)  # to coarse resolution
-        if self.A_up is not None:
-            self.A_up = self.A_up.to(x.device)
-            x_skip = self._truncate_fields(x_skip, self.A_up)  # back to high resolution
+        node_attributes_data = self.node_attributes(self._graph_name_data, batch_size=bse)
+        if grid_shard_shapes is not None:
+            # select grid_shard_slice of node_attributes but with proper gradient propagation in bwd
```
Comment (ssmmnn11) [inline models/src/anemoi/models/models/ens_encoder_processor_decoder.py]: same as above, can remove comment.

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 11]
```diff
@@ -239,6 +292,50 @@ def advance_input(
...
+        rollout_step: int,
+        training_mode: bool = True,
+        validation_mode: bool = False,
+    ) -> torch.Tensor:
+        is_sharded = self.grid_shard_slice is not None
+
+        sharding_supported = (self.loss_supports_sharding or not training_mode) and (
+            self.metrics_support_sharding or not validation_mode
+        )
+        if is_sharded and not sharding_supported:  # gather tensors if loss or metrics do not support sharding
```
Comment (ssmmnn11) [inline training/src/anemoi/training/train/forecaster/forecaster.py:308]: I wonder if we should fail here or print out a warning if the loss does not support sharding.

**Review by ssmmnn11 — CHANGES_REQUESTED** [review]
excellent :-)

some minor things

File: training/src/anemoi/training/train/forecaster/forecaster.py (hunk 11 above)
Comment (japols) [reply training/src/anemoi/training/train/forecaster/forecaster.py:308]: We print a warning in `__init__()` if either loss or val_metrics don't support sharding

**Review by ssmmnn11 — DISMISSED** [review]
very good :-)

**Review by ssmmnn11 — APPROVED** [review]
:-)
